spi/zynq7000-spi/libzynq7000-spi.c

```diff
@@ -0,0 +1,518 @@
+	*(spi->base + sr) = status;
+
+	/* TX FIFO empty */
+	if (status & (1 << 2)) {
+		*(spi->base + idr) = (1 << 2);
+		return spi->cond;
+	}
+
+	return 0;
+}
+
+
+#if 0
+static unsigned char spi_getMode(spi_t *spi)
+{
+	unsigned char mode = 0;
+	unsigned int cfg;
+
+	cfg = *(spi->base + cr);
+
+	if (cfg & (1 << 2)) {
+		mode |= SPI_CPHA;
+	}
+
+	if (cfg & (1 << 1)) {
+		mode |= SPI_CPOL;
+	}
+
+	return mode;
+}
+#endif
+
+
+static void spi_setMode(spi_t *spi, unsigned char mode)
+{
+	unsigned int cfg;
+
+	cfg = *(spi->base + cr);
+	cfg &= ~((1 << 2) | (1 << 1));
+
+	if (mode & SPI_CPHA) {
+		cfg |= (1 << 2);
+	}
+
+	if (mode & SPI_CPOL) {
+		cfg |= (1 << 1);
+	}
+
+	*(spi->base + cr) = cfg;
+}
+
+
+int spi_mode(unsigned int ndev, unsigned int ss, unsigned char mode)
+{
+	spi_t *spi;
+
+	spi = spi_get(ndev);
+	if (spi == NULL) {
+		return -ENODEV;
+	}
+
+	if (ss >= sizeof(spi->slaves) / sizeof(spi->slaves[0])) {
+		return -EINVAL;
+	}
+
+	spi->slaves[ss].mode = mode;
+
+	return EOK;
+}
+
+
+#if 0
+static unsigned int spi_getSpeed(spi_t *spi)
+{
+	unsigned int cfg;
+	unsigned char divisor;
+
+	cfg = *(spi->base + cr);
+	divisor = (cfg >> 3) & 7;
+
+	return SPI_SCLK(divisor);
+}
+#endif
+
+
+static void spi_setSpeed(spi_t *spi, unsigned int speed)
+{
+	unsigned int cfg;
+
+	cfg = *(spi->base + cr);
+	cfg &= ~(7 << 3);
+	cfg |= (__builtin_ffs(SPI_CLK_REF / speed) - 2) << 3;
+
+	*(spi->base + cr) = cfg;
+}
+
+
+int spi_speed(unsigned int ndev, unsigned int ss, unsigned int speed)
+{
+	spi_t *spi;
+	unsigned char divisor;
+
+	spi = spi_get(ndev);
+	if (spi == NULL) {
+		return -ENODEV;
+	}
+
+	if (ss >= sizeof(spi->slaves) / sizeof(spi->slaves[0])) {
+		return -EINVAL;
+	}
+
+	for (divisor = SPI_CLK_MIN_DIV; (divisor < SPI_CLK_MAX_DIV) && (SPI_SCLK(divisor) > speed); divisor++)
+		;
+
+	spi->slaves[ss].speed = SPI_SCLK(divisor);
+
+	return spi->slaves[ss].speed;
+}
+
+
+static void spi_select(spi_t *spi, int ss)
+{
+	unsigned int cfg;
+
+	if (ss < 0) {
+		/* Deselect all */
+		ss = 15;
+	}
+	else {
+		/* SS0 -> 0 */
+		/* SS1 -> 1 */
+		/* SS2 -> 3 */
+		if (ss == 2) {
+			ss++;
+		}
+	}
+
+	cfg = *(spi->base + cr);
+	cfg &= ~(15 << 10);
+	cfg |= (ss << 10);
+
+	*(spi->base + cr) = cfg;
+}
+
+
+int spi_xfer(unsigned int ndev, unsigned int ss, const void *out, void *in, size_t len)
+{
+	const unsigned char *obuff = out;
+	unsigned char data, *ibuff = in;
+	size_t n;
+	spi_t *spi;
+
+	spi = spi_get(ndev);
+	if (spi == NULL) {
+		return -ENODEV;
+	}
+
+	if (ss >= sizeof(spi->slaves) / sizeof(spi->slaves[0])) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return EOK;
+	}
+
+	/* Set slave SPI mode and speed */
+	spi_setMode(spi, spi->slaves[ss].mode);
+	spi_setSpeed(spi, spi->slaves[ss].speed);
+
+	/* Select slave and enable controller */
+	spi_select(spi, ss);
+	*(spi->base + er) = 1;
+
+	while (len > 0) {
+		/* Write to TX FIFO */
+		for (n = 0; (n < SPI_FIFO_SIZE) && (len > 0); len--, n++) {
+			/* TX FIFO full */
+			if (*(spi->base + sr) & (1 << 3)) {
```

Comment:
The `n < SPI_FIFO_SIZE` condition is necessary because the TX FIFO may never got full (the controller automatically removes data from TX FIFO). I've removed the `(*(spi->base + sr) & (1 << 3))` check - I agree that it's redundant.